### PR TITLE
Cartesian products introduced for 1D problems

### PR DESCRIPTION
**Pull request by mikaem**
## Summary

Extends `CartesianProduct` to support 1D mixed-variable formulations, in addition to the existing ND support. Previously `CartesianProduct` only accepted `TensorProductSpace` components; it now also accepts `OrthogonalSpace` / `DirectSum` components and returns the appropriate class.

Also consolidates `BlockTPMatrix` and the old `CartesianProductSpace` (ND-only) into a cleaner class hierarchy, and unifies the block-matrix infrastructure into a single `BlockMatrix` that handles both 1D and ND Cartesian product spaces.

## Types of Changes

- [x] Bug fix
- [x] New feature
- [x] Refactor / cleanup

## Related Issues

Refs #49

## Description of Changes

**Class hierarchy**

- `CartesianBaseSpace` (new abstract base) holds all shared logic (`evaluate_mesh`, `backward`, `forward`, `scalar_product`, `to_orthogonal`, `from_orthogonal`, `block_sizes`, `dim`, `shape`, `num_dofs`).
- `CartesianTensorProductSpace` (renamed from `CartesianProductSpace`) — ND composite, components are `TensorProductSpace` / `DirectSumTPS`.
- `CartesianProductSpace` (new) — 1D composite, components are `OrthogonalSpace` / `DirectSum`. Overrides `backward`, `evaluate_mesh`, and `backward_primitive` with scalar `N: int | None` signatures appropriate for 1D spaces.
- `VectorTensorProductSpace` now inherits from `CartesianTensorProductSpace`.
- `CartesianProduct` factory dispatches based on `dims`: all-1D → `CartesianProductSpace`, all-ND → `CartesianTensorProductSpace` (or `VectorTensorProductSpace` for `rank=1`). Mixed dims are rejected with `AssertionError`.

**Block matrix infrastructure**

- `BlockTPMatrix` removed; replaced by a single generic `BlockMatrix[MatrixT, SpaceT]` that handles both `TPMatrix`-based (ND) and `GlobalMatrix`-based (1D) blocks.
- `IndexedArray` removed; replaced by `GlobalArray` and `GlobalMatrix` (in `la/operators.py`), which carry a `global_indices` tag used to route contributions into the correct block during assembly.
- `BlockMatrix.tosparse()` assembles a global `DiaMatrix` by placing each block at the correct diagonal offset, with result caching.

**1D inner product assembly**

- `inner()` over a `CartesianProductSpace` now produces a `BlockMatrix` of `GlobalMatrix` blocks (bilinear) or a `BlockArray` (linear), routed by `global_index` on each component space.
- `OrthogonalSpace` and `DirectSum` gain `leaf` and `global_index` attributes set during `CartesianProductSpace.__init__`.

**Other**

- `TensorProductSpace.shape` converted from method to `@property`; now returns `space.num_quad_points` per axis (not raw mode count `N`).
- `OrthogonalSpace` gains `to_orthogonal` / `from_orthogonal` identity methods and `shape` property for API symmetry with `Composite`.
- `Composite` and `DirectSum` gain `evaluate_mesh` methods.
- Example `examples/coupled1D.py` added, demonstrating a 1D mixed Poisson formulation.

## Breaking Changes

- [x] Yes

`CartesianProductSpace` is now the **1D** class. The previous ND class is now `CartesianTensorProductSpace`. Code that imports or does `isinstance(..., CartesianProductSpace)` for ND spaces must be updated to use `CartesianTensorProductSpace`.

`BlockTPMatrix` is removed; replace with `BlockMatrix`. `IndexedArray` is removed; replace with `GlobalArray`.

`TensorProductSpace.shape` is now a `@property` (previously a method); callers using `T.shape()` must drop the parentheses.

## Checklist

- [x] Tests added or updated
- [x] Documentation updated (README, examples, docstrings)
- [x] Added type hints
- [x] Linting & formatting pass locally (`pre-commit run --all-files`)
- [x] All new and existing tests pass (`uv run pytest`)
- [ ] Coverage does not decrease
- [ ] Git history is clean (squash/fixup before merge)

## Additional Notes

N/A

### CODE REVIEW

File: src/jaxfun/galerkin/cartesianproductspace.py [hunk 1]
```diff
@@ -1,60 +1,221 @@
...
 from jaxfun.sharding import physical_sharding, spectral_sharding
 from jaxfun.typing import MeshKind
 
+type MultiDimensionalSpace = TensorProductSpace | CartesianTensorProductSpace
+type OneDimensionalSpace = OrthogonalSpace | DirectSum | CartesianProductSpace
 
+
+@overload
+def CartesianProduct(
+    *basespaces: MultiDimensionalSpace, name: str = "CP", rank: int = -1
```
Comment (augustfe) [inline src/jaxfun/galerkin/cartesianproductspace.py]: Change `rank` to enum?

File: src/jaxfun/la/blockmatrix.py [hunk 2]
```diff
@@ -1,65 +1,90 @@
...
-                _grouped[idx] = TPMatrices([mat])
-            else:
-                _grouped[idx].tpmats.append(mat)
-        self._combined_blocks = nnx.Dict(
-            {idx: tpmats_to_kron(list(tpm.tpmats)) for idx, tpm in _grouped.items()}
-        )
+        if isinstance(mats[0], TPMatrix):
+            _grouped: dict[str, TPMatrices] = {}
+            for mat in mats:
+                idx = f"{mat.global_indices[0]},{mat.global_indices[1]}"
```
Comment (augustfe) [inline src/jaxfun/la/blockmatrix.py:63]: Why use a string here instead of just a tuple of ints directly?

File: src/jaxfun/la/blockmatrix.py [hunk 3]
```diff
@@ -1,65 +1,90 @@
...
+                if key not in _grouped:
+                    _grouped[key] = cast(DiaMatrix | Matrix, mi.data)
+                else:
+                    _grouped[key] = _grouped[key] + mi.data
+            self._combined_blocks = nnx.Dict(_grouped)
+
+    def _matmul_array(self, w: tuple[Array, ...]) -> tuple[Array, ...]:
+        shapes = self.test_space.num_dofs
+        out = [jnp.zeros(shape) for shape in shapes]
+        for s, block_mat in self._combined_blocks.items():
```
Comment (augustfe) [inline src/jaxfun/la/blockmatrix.py:85]: I think this should be able to be restructured so that we don't loop over everything, vectorizing over the indices and maybe `tree.map` over the block matrices?

File: src/jaxfun/la/operators.py [hunk 4]
```diff
@@ -244,3 +250,24 @@ def __rsub__[T](self, other: T) -> T:
 
     def __getitem__(self, key: tuple[int, int], /) -> Array:
         return jnp.zeros((), dtype=self.dtype)
+
+
+class GlobalMatrix(BaseMatrix):
```
Comment (augustfe) [inline src/jaxfun/la/operators.py]: This looks more like it should be a MixIn maybe, so that you don't have to do `matrix.data` everywhere to do work

File: src/jaxfun/la/blockmatrix.py (hunk 2 above)
Comment (mikaem) [reply src/jaxfun/la/blockmatrix.py:63]: It's because it is later stored as an nnx.Dict, which requires str keys.

File: src/jaxfun/la/blockmatrix.py (hunk 3 above)
Comment (mikaem) [reply src/jaxfun/la/blockmatrix.py:85]: Possibly. But there are very few blocks, so may not be worth it.

File: src/jaxfun/la/operators.py (hunk 4 above)
Comment (mikaem) [reply src/jaxfun/la/operators.py]: Perhaps, but I don't see how? The MixIn would require its own state (global indices) and `__init__`, which is not really MixIn?

Comment (mikaem) [reply src/jaxfun/la/operators.py]: There is one possible solution @augustfe, but not sure it's the best? At least now GlobalMatrix has all the BaseMatrix methods:

```
class _DataDelegatingMixin:
    """Mixin that delegates all BaseMatrix operations to ``self.data``."""

    data: DiaMatrix | Matrix

    @property
    def shape(self) -> tuple[int, int]:
        return self.data.shape

    @property
    def dtype(self) -> jnp.dtype:
        return self.data.dtype

    def matvec(self, x: Array, axis: int = 0) -> Array:
        return self.data.matvec(x, axis=axis)

    def lu_factor(self):
        return self.data.lu_factor()

    def lu_solve(self, b: Array, axis: int = 0) -> Array:
        return self.data.lu_solve(b, axis=axis)

    def solve(self, b: Array, axis: int = 0) -> Array:
        return self.data.solve(b, axis=axis)

    @property
    def T(self) -> BaseMatrix:
        return self.data.T

    def diagonal(self, k: int = 0) -> Array:
        return self.data.diagonal(k)

    def diagonal_or_none(self) -> Array | None:
        return self.data.diagonal_or_none()

    @property
    def is_diagonal(self) -> bool:
        return self.data.is_diagonal

    def todense(self) -> Array:
        return self.data.todense()

    def tosparse(self, *, tol: int = 100) -> DiaMatrix:
        return self.data.tosparse(tol=tol)

    def to_matrix(self) -> Matrix:
        return self.data.to_matrix()

    def get_row(self, i: int | Array) -> Array:
        return self.data.get_row(i)

    def get_column(self, j: int | Array) -> Array:
        return self.data.get_column(j)

    @property
    def size(self) -> int:
        return self.data.size

    def __call__(self, u: Array | JAXFunction) -> Array:
        return self.data(u)

    def __getitem__(self, key: tuple[int, int], /) -> Array:
        return self.data[key]

    def __matmul__(self, other) -> Array:
        return self.data @ other

    def __rmatmul__(self, other) -> Array:
        return other @ self.data

    def __add__(self, other):
        return self.data + other

    def __len__(self) -> int:
        return len(self.data)


class GlobalMatrix(_DataDelegatingMixin, BaseMatrix):
    data: DiaMatrix | Matrix

    def __init__(self, global_indices: tuple[int, int], matrix: DiaMatrix | Matrix):
        self.data = matrix
        self.global_indices = global_indices

    def scale(self, alpha: complex | Array) -> GlobalMatrix:
        return GlobalMatrix(self.global_indices, self.data.scale(alpha))

    def astype(self, dtype: jnp.dtype) -> GlobalMatrix:
        return GlobalMatrix(self.global_indices, self.data.astype(dtype))

```

Comment (mikaem) [reply src/jaxfun/la/operators.py]: Reconsidering @augustfe and I think this is overcomplicating stuff. The GlobalMatrix just needs two global indices such that BlockMatrix can be created from a list[GlobalMatrix]. So GlobalMatrix does not need all the methods of BaseMatrix and it can be implemented simply as

```
class GlobalMatrix(nnx.Pytree):
    def __init__(self, global_indices: tuple[int, int], matrix: DiaMatrix | Matrix):
        self.data = nnx.data(matrix)
        self.global_indices = nnx.static(global_indices)

    def scale(self, alpha: complex | Array) -> GlobalMatrix:
        return GlobalMatrix(self.global_indices, self.data.scale(alpha))

```
The scale is used on BlockMatrix.scale. Otherwise consistent with GlobalArray, implemented as 

```
class GlobalArray(nnx.Pytree):
    def __init__(self, i: int, data: Array):
        self.data = data
        self.i = i
```
